### ISSUE TITLE
[HOLD for payment 2023-11-09] [$500] mWeb/Safari- Chat - After tapping on "Copy to clipboard", options menu does not close

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#28820]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any chat
2. Long press on any message
3. Tap on "Copy to clipboard"

## Expected Result:
There should be a message indicating the message has been copied and then the options menu should close

## Actual Result:
There is no message indicating that the message has been copied and options menu does not close

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.77-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/3c7d3265-135e-46ed-89f6-6da4ac595a8b

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0153aad7aa52da20d4</li>
        <li>Upwork Job ID: 1710335830550470656</li>
        <li>Last Price Increase: 2023-10-13</li>
<li>Automatic offers: </li>
<ul>
<li>cubuspl42 | Reviewer | 27277511</li>
<li>AmjedNazzal | Contributor | 27277514</li>
</ul></li>
    </ul>
</details>

**Comment by paultsimura** [comment]
Not reproducible on Staging and latest `main`

**Comment by garrettmknight** [comment]
I was able to reproduce on staging and prod - It's not the "+" button, it's a long press on any message in a DM that gets you to the options menu for the chat on mWeb.

**Comment by ChrisCates** [comment]
The best option would be to update the menu to use touch events.
More specifically the `touchend` event.

Review the documentation on Mozilla as seen [here](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events/Using_Touch_Events).

Would be more than happy to submit a PR. Let me know if I can work on the project.

**Comment by cubuspl42** [comment]
@ChrisCates

You're welcome to submit a proposal! Please [read the contributing guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md).

Also, take a quick look at any other issue to see how the proposal selection process looks in Expensify.

**Comment by ChrisCates** [comment]
@cubuspl42 I'll prepare a proposal shortly.
I've also not been invited to Slack yet.

Should I wait for that? Or can I submit PRs and everything will be cool?

**Comment by ChrisCates** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue

There is an issue with pointer events on mobile for Safari with the chat menu. More specifically the copy to clipboard button is not working properly. We should verify and test all buttons though.

### What is the root cause of that problem?

After further review, we should do regression testing on all menu items that utilize:

```
src/components/CopyTextToClipboard.js
src/components/Pressable/PressableWithDelayToggle.js
src/components/Pressable/GenericPressable/...
```

### What changes do you think we should make in order to solve the problem?

We should investigate an implementation that uses `Touch` events. We will need to implement this in tandem with existing Pressable This way, if there is a long press, it will still register with the `touchend` event.

The following is Mozilla Documentation on Touch Events and how to implement them: https://developer.mozilla.org/en-US/docs/Web/API/Touch_events/Using_Touch_Events

We would also need to verify on all web platforms that the new touch events work as intended and do not have any potential negative side effects. We should verify mobile on React Native and Desktop App too.

**Comment by cubuspl42** [comment]
@ChrisCates 

> Should I wait for that? Or can I submit PRs and everything will be cool?

From [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md):

> Refrain from leaving additional comments until someone from the Contributor-Plus team and / or someone from Expensify provides feedback on your proposal (do not create a pull request yet)

**Comment by cubuspl42** [comment]
@ChrisCates Congrats on writing your first proposal! As is often the case with first tries, there's a big room for improvement here. Both crucial sections of the proposal (the root cause analysis and the solution description) are not  up to Expensify's standards.

> What is the root cause of that problem?
> 
> After further review, we should do regression testing 

How does "After further review, we should do regression testing..." answer the question "What is the root cause of that problem?"?

> What changes do you think we should make in order to solve the problem?

> We should investigate...

The investigation is what we typically do when we prepare a proposal. The proposal itself should contain a description of the solution itself. Depending on the scope of the issue, the level of detail varies. For relatively small issues (the majority), it is typical to have the full solution prepared. The proposal then contains a description of that solution.

_Please_ take a look at a dozen other issues to get a feeling of how the Expensify process works. The crucial labels are "External" and "Help wanted". Issues with "Help wanted" are open for new proposals, while the issues without "Help wanted" are typically already being handled.

[Here are the "External" issues without "Help wanted"](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+sort%3Aupdated-desc+label%3AExternal+-label%3A%22Help+Wanted%22) (i.e. mostly the ones that are being handled). Check out the proposals that were selected for inspiration.

**Comment by ChrisCates** [comment]
Hey @cubuspl42 thanks for the feedback. Just came here.
I'll make sure to just add comments and feedback next time. Also I think @ishpaul777 has the better solution. Since, to be honest, I was struggling to reproduce on LAN.

I have a question about this issue. Is this more in line with what follows Expensify's guidelines. And if not? What should change: #28078

**Comment by cubuspl42** [comment]
@ishpaul777 Please stick to the proposal template. The headers are not meant to be modified. Thank you!

**Comment by cubuspl42** [comment]
@lanitochka17 @garrettmknight I can't reproduce on staging

**Comment by garrettmknight** [comment]
@cubuspl42 I just reproduced it on staging as well on 1.3.81-6

**Comment by cubuspl42** [comment]
@ishpaul777

> selection.setBaseAndExtent throws a type error

Are there any details available for this error?

> we can have a check for anchorNode focusNode to be not null.

Have you analyzed the surrounding code? Do we understand when the mentioned values could be `null`?

**Comment by cubuspl42** [comment]
> Personally i can still reproduce in mobile and desktop both.

Would you name the specific browsers / platforms? This issue is labeled mWeb/Safari, so please let us know if you found this to be a wrong labeling

**Comment by cubuspl42** [comment]
I cannot reproduce this on my physical iPhone, neither on staging nor on prod. We definitely need to figure things out here.

@lanitochka17 Would you verify if this is still reproducible for you? If so, on what device?

**Comment by lanitochka17** [comment]
@cubuspl42 Hello 
Issue reproducible on the latest Build 1.3.85-2
Iphone 14 Pro Max/Safari


http://github.com/Expensify/App/assets/78819774/f5c28956-59e2-4e64-93e5-29a24efc797f

**Comment by cubuspl42** [comment]
@ishpaul777 

> The issue can only be spotted on production and staging and on emulator(dev) for me

So you can observe this on production and staging also on a physical iPhone, did I get this right?

**Comment by cubuspl42** [comment]
@ishpaul777 Thanks for these detail!

Still, I'd like to see a proposal with a deeper root cause analysis...

I [advertised the issue on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1697616345952899).

**Comment by AmjedNazzal** [comment]
I can't reproduce this at all, not in staging, not in prod, so unless this is phone model specific I don't know what it is because I'm trying on simulated iphone 14 and the options menu goes away every time.

**Comment by AmjedNazzal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

mWeb/Safari- Chat - After tapping on "Copy to clipboard", options menu does not close

### What is the root cause of that problem?

Initially I could not reproduce this bug on MacOS Safari nor on iOS Safari, but turned out the reason I could not reproduce it on iOS is because I was using a simulated iPhone 14 which doesn't have the recent versions of Safari and upon updating Safari on my MacOS I could reproduce it, this points to the cause being recent or relatively recent changes in WebKit code 

Looking at the current code from WebKit:

https://github.com/WebKit/WebKit/blob/29de3632a0deb774e13bed1d197aa953a2bff7ae/Source/WebCore/page/DOMSelection.cpp#L292-L293

It seem like we can no longer pass null to setBaseAndExtent, I don't know when this was added or how long this bug here has been going on unnoticed but we clearly cannot provide null here and initial selection.getSelection() in our code is providing null for anchorNode since there was no user selection to begin with, this anchorNode we are passing to setBaseAndExtent as the argument that should be called baseNode instead of anchorNode at the moment, which in itself doesn't matter but it means that setBaseAndExtent will understand that baseNode is null and same with focusNode which is now extentNode in setBaseAndExtent, so to summarize, setBaseAndExtent no longer takes anchorNode and focusNode as null but we are passing null for both.

It seem like WebKit is trying to remove the use of null overall, if you look at the following recent PR they seem to want to always use container for nodes instead of null:

https://github.com/WebKit/WebKit/commit/457eb4252b1fe6cf602a8cd1a7ceef05195e579b

### What changes do you think we should make in order to solve the problem?

We should only use setBaseAndExtent if we do have an anchorNode that wasn't cleared/reset

```javascript
else if(originalSelection.anchorNode){
    selection.setBaseAndExtent(originalSelection.anchorNode, originalSelection.anchorOffset, originalSelection.focusNode, originalSelection.focusOffset);
}
```

Testing the code, for the recent Safari, selection is reset to empty after selection.removeAllRanges(); so I don't think we need to further use setBaseAndExtent 

### Result

https://github.com/Expensify/App/assets/74202751/0082cac2-b72b-4742-a9a8-9c6d42a6a96a

**Comment by AmjedNazzal** [comment]
@cubuspl42 Ahhh that makes even more sense now, thanks for the further analysis! I agree with their decision and think null shouldn't be a passable argument for something like this.

**Comment by cubuspl42** [comment]
@ishpaul777 

It's because I disapproved of the approach you took with your proposal.

> selection.setBaseAndExtent throws a type error when "Copy to Clipboard" is clicked.

This is merely an observation

>  I am still unsure why this fails for some devices and not for others

This shows that you didn't find it crucial to understand the phenomena you're providing a solution for

> Or we use try, catch like below

This again indicates that you're not firmly interested in understanding the problem; also, try-catch-swallow is a bad pattern in software engineering and, frankly, suggesting it (even as an alternative) added negative value to the proposal

> You can try open this example in safari latest version and replace the arguments passed to setBaseAndExtent as null, theres a good chance you will get the typeerror

This is an incentive to test this with a note that some result is _likely_ to be observed

While the proposal by @AmjedNazzal had some flaws (as I noted, it would be better if the specs were checked), it showed an active interest in understanding the problem and the solution, at least by providing links to the WebKit implementation.

To sum up, the Expensify proposal template is split into three crucial sections for a good reason. They all have to provide value, while in your case only the first (problem summary) and the last one (solution) did.

I hope that this feedback will help you write better proposals in the future and I'm sorry if my decision is disappointing for you.

**Comment by AmjedNazzal** [comment]
@ishpaul777 I don't think you can call this biased, there is no incentive for @cubuspl42 to accept one proposal over the other aside from the proposal itself, I think his issue with your was the lack of explanation of why this is happening when it used to work, it's important to understand why something changed because otherwise we can never be sure if a certain fix will last or if the issue will occur again, understanding that the reason this started happening is because of the version of safari and the changes in WebKit repo is very important because otherwise it's just a mysterious thing that happened for unknown reasons.

**Comment by cubuspl42** [comment]
I don't know what I'm biased on and I'm not sure if I want to know. I explained as well as I could why I disapproved of your proposal even if the solution itself turned out to be okay. It's natural that the author considers their proposal the best, and it's also natural that the reviewer sometimes disagrees. It happened to me multiple times and it will probably happen to you again if you decide to contribute to Expensify in the future.

**Comment by cubuspl42** [comment]
@tgolen (the assigned internal engineer) can comment on this situation if they choose to do so

**Comment by tgolen** [comment]
Giving @AmjedNazzal the 🟢 for their proposal and assigning.

**Comment by AmjedNazzal** [comment]
@cubuspl42 PR is ready for review. Sorry it took so long, I had some pc issues.

**Comment by cubuspl42** [comment]
@tgolen Ouch, is it the GitHub autoclosing based on the commit message? 😳

**Comment by tgolen** [comment]
Oh, hahaha, yes. I'll reopen this.

**Comment by cubuspl42** [comment]
@tgolen Do we actually use that behavior? If we don't, maybe it should be disabled in the settings?

**Comment by tgolen** [comment]
I don't think we use it in this repo, no. Do you know where the setting might be? I'm looking at another repo I own and I don't know where to find it.

**Comment by cubuspl42** [comment]
@tgolen I might've confused it with GitLab. It appears that on GitHub [it cannot be disabled](https://github.com/orgs/community/discussions/17308) 😕

It could be worth adding a PR-level check to prevent [`close|closes|closed|fixes|fixed #\d+`](https://github.blog/2013-01-22-closing-issues-via-commit-messages/)

**Comment by tgolen** [comment]
Ah, I agree. I confirmed with our infra team that they don't know of a setting like that either. I think we're OK for now. It doesn't happen very often.

**Comment by cubuspl42** [comment]
> ⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/30796)

False alarm.

**Comment by cubuspl42** [comment]
- The PR that introduced the bug has been identified. Link to the PR:
    -  N/A; this is not a regression, but an adjustment to the changed behavior of the new Safari version
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
    - N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    - No need for additional discussion
- Determine if we should create a regression test for this bug.
    - Up to the QA team
- If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
    - Launch the application on macOS Safari 17 or higher or Safari on iPhone 15 or higher.
    - Navigate to any chat and right-click or long-press on a chat message to display the context menu.
    - Select 'Copy to clipboard'.
    - Ensure the text is copied successfully (verify this by pasting it somewhere) and the context menu closes after the action.

**Comment by AmjedNazzal** [comment]
@garrettmknight Gentle nudge for the payment. Thanks! :)

**Comment by garrettmknight** [comment]
Summary of payments:
- Urgency bonus: No
- Contributor: @AmjedNazzal $500 via Upwork
- Reviewer: @cubuspl42 $500 via Upwork

**Comment by garrettmknight** [comment]
Since this is really version specific and a pretty edge case, I don't think we need a regression test.
